examples/form/src/components/preact/Form.tsx

```diff
@@ -0,0 +1,125 @@
+	toTrackAstroSubmitStatus,
+	toValidateField,
+	validateForm,
+	formNameInputProps,
+} from "simple:form";
+
+export function useCreateFormContext(
+	validator: FormValidator,
+	fieldErrors?: FieldErrors,
+) {
+	const initial = getInitialFormState({ validator, fieldErrors });
+	const [formState, setFormState] = useState<FormState>(initial);
+	return {
+		value: formState,
+		set: setFormState,
+		setValidationErrors: toSetValidationErrors(setFormState),
+		validateField: toValidateField(setFormState),
+		trackAstroSubmitStatus: toTrackAstroSubmitStatus(setFormState),
+	};
+}
+
+export function useFormContext() {
+	const formContext = useContext(FormContext);
+	if (!formContext) {
+		throw new Error(
+			"Form context not found. `useFormContext()` should only be called from children of a <Form> component.",
+		);
+	}
+	return formContext;
+}
+
+type FormContextType = ReturnType<typeof useCreateFormContext>;
+
+const FormContext = createContext<FormContextType | undefined>(undefined);
+
+export function Form({
+	children,
+	validator,
+	context,
+	fieldErrors,
+	name,
+	...formProps
+}: {
+	validator: FormValidator;
+	context?: FormContextType;
+	fieldErrors?: FieldErrors;
+} & Omit<ComponentProps<"form">, "method" | "onSubmit">) {
+	const formContext = context ?? useCreateFormContext(validator, fieldErrors);
+
+	return (
+		<FormContext.Provider value={formContext}>
+			<form
+				{...formProps}
+				method="POST"
+				onSubmit={async (e) => {
+					e.preventDefault();
+					const formData = new FormData(e.currentTarget);
+					formContext.set((formState) => ({
+						...formState,
+						isSubmitPending: true,
+						submitStatus: "validating",
+					}));
+					const parsed = await validateForm({ formData, validator });
+					if (parsed.data) {
+						return formContext.trackAstroSubmitStatus();
+					}
+
+					e.stopPropagation();
+					formContext.setValidationErrors(parsed.fieldErrors);
+				}}
+			>
+				{name ? <input {...formNameInputProps} value={name} /> : null}
+				{children}
+			</form>
+		</FormContext.Provider>
+	);
+}
+
+export function Input(inputProps: ComponentProps<"input"> & { name: string }) {
+	const formContext = useFormContext();
+	const fieldState = formContext.value.fields[inputProps.name];
+	if (!fieldState) {
+		throw new Error(
+			`Input "${inputProps.name}" not found in form. Did you use the <Form> component?`,
+		);
+	}
+
+	const { hasErroredOnce, validationErrors, validator } = fieldState;
+	return (
+		<>
+			<input
+				onBlur={async (e) => {
+					// @ts-ignore
+					const value: string = e.target.value; // ? this is saying that e.target can be null, and value doesn't exist on EventTarget
+					if (value === "") return;
+					formContext.validateField(inputProps.name, value, validator);
+				}}
+				onInput={async (e) => {
+					if (!hasErroredOnce) return;
+					// @ts-ignore
+					const value: string = e.target.value; // ? this is saying that e.target can be null, and value doesn't exist on EventTarget
```

Comment:
```suggestion
				onBlur={async (e) => {
					const value = e.currentTarget.value;
					if (value === "") return;
					formContext.validateField(inputProps.name, value, validator);
				}}
				onInput={async (e) => {
					if (!hasErroredOnce) return;
					const value = e.currentTarget.value;
```

Ah, looks like Preact is more strict about `value`! We can use `currentTarget` to guarantee the event is using an input element. Checked locally that this fixes our type issue 👍